time_util.py

```diff
@@ -400,3 +401,149 @@ def get_frame_count(start, end, td):
 
 def is_archaeological_layer(layer):
     return layer.getTimeFormat() in [bcdate_util.BC_FORMAT]
+
+
+def to_discrete_datetime(extent, time_type, time_frame):
+    """
+    :param extent: current start end end time (as (datetime, datetime) tuple) from project extents
+    :param time_type: one of "minutes", "years" etc
+    :param time_frame: length of a time frame in the TimeManager plugin (in time_type units!)
+    :return: an extent, being a tuple of a start datetiem and an end datetime
+
+    Given the time_frame, time_type and time_start, return a new start/endtime in which current start_time fits,
+    but is starting on a time_start (of type time_type) and within not more then one time_frame from current
+    end_time.
+
+    In other words: find the first (last) best discrete datetime before current start_time and best discrete
+    endtime after current end_time
+    """
+
+    dt0 = extent[0]  # datetime 0 /start
+    d0 = dt0.date()  # date 0 /start
+    t0 = dt0.time()  # time 0 / start
+    # dte, de and te stand for datetime-end, date-end, time-end
+
+    if time_type == "microseconds":
```

Comment:
Important: this can be made much more compact without repetitions if you pass kwargs to all the functions that require a varying kw arg (ie microseconds, minutes etc).